### PR TITLE
Common type between character types should be another character type.

### PR DESCRIPTION
**Pull request by quickfur**
This is my hamfisted fix for [issue 17141](https://issues.dlang.org/show_bug.cgi?id=17141). I'm pretty sure it breaks something out there, but at least it worked for the use case described in the bug! (The reason for submitting it here is really so that the autotester can tell me what I've broken.)

### CODE REVIEW

**Comment by JackStouffer** [comment]
Looks like a lot of the dub packages aren't being built. Unfortunately it doesn't give much in the way of debug info.

**Comment by quickfur** [comment]
The autotester shows a backend ICE and a problem with implicit conversions to/from integer types. I think this is beyond my knowledge of dmd internals to fix. :-(

**Comment by JackStouffer** [comment]
Might as well just special case Phobos then :/

**Comment by quickfur** [comment]
I don't think it's a good idea to special case Phobos... because you'll end up with the same problem when you use `?:` directly with char types.

File: src/ddmd/dcast.d [hunk 1]
```diff
@@ -2701,7 +2701,7 @@ extern (C++) bool typeMerge(Scope* sc, TOK op, Type* pt, Expression* pe1, Expres
     Type t1b = e1.type.toBasetype();
     Type t2b = e2.type.toBasetype();
 
-    if (op != TOKquestion || t1b.ty != t2b.ty && (t1b.isTypeBasic() && t2b.isTypeBasic()))
+    if (op != TOKquestion && t1b.ty != t2b.ty && (t1b.isTypeBasic() && t2b.isTypeBasic()))
```
Comment (ibuclaw) [inline src/ddmd/dcast.d]: You probably don't want to have `&&` here.

**Comment by ibuclaw** [comment]
I think it is correct that two different types are promoted to the smallest common type starting from  `int`.

**Comment by JackStouffer** [comment]
@ibuclaw this code

```d
import std.stdio;

void main()
{
	char a = 'a';
	dchar b = 'b';
	auto c = true ? a : b;
	writeln(c);
}
```

prints

```
97
```

which is nonsense.

**Pull request closed without merging**